src/application-manager.coffee

```diff
@@ -0,0 +1,1047 @@
+					@services.handover(step.current, step.target)
+			fetch: (step) =>
+				startTime = process.hrtime()
+				@fetchesInProgress += 1
+				Promise.join(
+					@config.get('fetchOptions')
+					@images.getAvailable()
+					(opts, availableImages) =>
+						opts.deltaSource = @bestDeltaSource(step.image, availableImages)
+						@images.triggerFetch step.image, opts, (success) =>
+							@fetchesInProgress -= 1
+							@timeSpentFetching += process.hrtime(startTime)[0]
+							if success
+								@reportCurrentState(update_downloaded: true)
+				)
+			removeImage: (step) =>
+				@images.remove(step.image)
+			saveImage: (step) =>
+				@images.save(step.image)
+			cleanup: (step) =>
+				@images.cleanup()
+			createNetworkOrVolume: (step) =>
+				model = if step.model is 'volume' then @volumes else @networks
+				model.create(step.target)
+			removeNetworkOrVolume: (step) =>
+				model = if step.model is 'volume' then @volumes else @networks
+				model.remove(step.current)
+			ensureSupervisorNetwork: =>
+				@networks.ensureSupervisorNetwork()
+		}
+		@validActions = _.keys(@actionExecutors).concat(@proxyvisor.validActions)
+		@_router = new ApplicationManagerRouter(this)
+		@router = @_router.router
+
+		@images.on('change', @reportCurrentState)
+		@services.on('change', @reportCurrentState)
+
+	serviceAction: serviceAction
+	imageForService: imageForService
+	fetchAction: fetchAction
+
+	reportCurrentState: (data) =>
+		@emit('change', data)
+
+	init: =>
+		@images.cleanupDatabase()
+		.then =>
+			@services.attachToRunning()
+		.then =>
+			@services.listenToEvents()
+
+	# Returns the status of applications and their services
+	# TODO: discuss: I think commit could be deduced by the UI looking at the image_installs on the API?
+	getStatus: =>
+		Promise.join(
+			@services.getStatus()
+			@images.getStatus()
+			@db.models('app').select([ 'appId', 'releaseId', 'commit' ])
+			(services, images, targetApps) ->
+				apps = {}
+				dependent = {}
+				releaseId = null
+				# We iterate over the current running services and add them to the current state
+				# of the app they belong to.
+				for service in services
+					appId = service.appId
+					apps[appId] ?= {}
+					apps[appId].services ?= {}
+					# We only send commit if all services have the same release, and it matches the target release
+					if !releaseId?
+						releaseId = service.releaseId
+					else if releaseId != service.releaseId
+						releaseId = false
+					if !apps[appId].services[service.imageId]?
+						apps[appId].services[service.imageId] = _.pick(service, [ 'status', 'releaseId' ])
+						apps[appId].services[service.imageId].download_progress = null
+					else
+						# There's two containers with the same imageId, so this has to be a handover
+						previousReleaseId = apps[appId].services[service.imageId].releaseId
+						apps[appId].services[service.imageId].releaseId = Math.max(previousReleaseId, service.releaseId)
```

Comment:
I don't think that this logic is fully correct, with release pinning.

Reply:
Changed to use the releaseId of the oldest container